### ISSUE TITLE
[Clang][C++23] [[assume]] falsely rejects trivial side-effect-free function calls as having side-effects

### ISSUE THREAD

**Issue by SekaiArendelle** [llvm/llvm-project#194242]
## Description

Clang's current implementation of C++23 [[assume(expr)]] conservatively treats all function calls as having side effects. This is because Sema::BuildCXXAssumeExpr uses Expr::HasSideEffects(Context), which does not inspect function bodies and simply flags every CallExpr as potentially impure.
This produces false-positive -Wassume warnings and causes the assumption to be ignored entirely, even when the callee is a trivial, pure helper function whose body is visible and provably side-effect free.

## Minimal Reproducer
```cpp
// compile with: clang++ -std=c++23
struct X {
    bool foo() const {
        return true;
    }
};

int main() {
    X x{};
    [[assume(x.foo())]];           // false-positive warning
    [[assume(x.foo() == true)]];   // false-positive warning
}
```

## Actual Behavior
```
warning: assumption is ignored because it contains (potential) side-effects [-Wassume]
   10 |         [[assume(x.foo())]];
      |         ~~~~~~~~~^~~~~~~~~~
warning: assumption is ignored because it contains (potential) side-effects [-Wassume]
   11 |         [[assume(x.foo() == true)]];
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~~
```

## Expected Behavior
No warning. The assumption should be retained because X::foo() is trivially side-effect free.

## Real-world impact
This over-approximation makes [[assume]] unusable for many common defensive-programming patterns, such as:

```
[[assume(i < vec.size())]];        // size() is typically a pure accessor
[[assume(is_valid_index(i))]];     // user-defined pure helpers
```
All of these are currently rejected, forcing developers to either suppress the warning or avoid [[assume]] entirely, undermining the feature's utility.

## Environment
clang version: trunk
Standard: -std=c++23 or -std=c++26

**Comment by shafik** [comment]
Linking this previous discussion on assumptions and optimization issues: https://discourse.llvm.org/t/llvm-assume-blocks-optimization/71609 

See more detailed discussion on issues in the PR linked above on why this is currently something we won't fix.
